<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
	<head>
	   <title>Report NA12878_03</title>
	   <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
	   <style type="text/css">
		<!--
body
{
	font-family: Calibri, sans-serif;
	font-size: 8pt;
}
h4
{
	font-family: Calibri, sans-serif;
	font-size: 10pt;
}
table
{
	border-collapse: collapse;
	border: 1px solid black;
	width: 100%;
}
th, td
{
	border: 1px solid black;
	font-size: 8pt;
	text-align: left;
}
p
{
 margin-bottom: 0cm;
}
		-->
	   </style>
	</head>
	<body>
<h4>Technical Report for Bioinformatic Analysis</h4>

<p><b>Sample: NA12878_03</b> (ex1)
<br />
<br />sample sex: female
<br />Processing system: HaloPlex HSP v2
<br />Processing system type: Panel Haloplex
<br />Reference genome: GRCh37
<br />Date: 19.02.2021
<br />Analysis pipeline: megSAP 0.2-486-gc928f885
<br />Analysis software: cppNGSD-TEST 
</p>

<p><b>Phenotype information</b>
</p>

<p><b>Target region</b>
<br /><span style="font-size: 8pt;">The target region includes CCDS ("consensus coding sequence") of the genes listed below &plusmn;20 flanking bases of the intronic sequence. It may comprise additional exons and/or flanking bases.
<br />Name: panel
<br />Genes analyzed: 3 (see coverage statistics)
</span></p>

<p><b>Criteria for variant filtering </b>
<br />&nbsp;&nbsp;&nbsp;&nbsp;- Allele frequency &le; 1.00%
<br /><br />Small variants in target region: 135
<br />SNVs/InDels selected for report: 2
<br />CNVs/SVs selected for report: 5
</p>
<br />If present, the following tables contain: likely pathogenic variants (class 4)<sup>*</sup> and pathogenic variants (class 5)<sup>*</sup>, for which a contribution to the clinical symptoms of the patient is conceivable, and variants of uncertain significance (class 3)<sup>*</sup>, for which a further evaluation of the clinical relevance by follow-up examinations may be useful.
Depending on the type of genetic alteration, family history and clinical features of the patient further investigations might change the classification of variants.
<br />A (uncommented) list of all detected variants can be provided on request.
<br />In case of a suspected clinical diagnosis genetic counseling is necessary to evaluate the indication/possibility of further genetic studies.
<br /><sup>*</sup> For information on the classification of variants, see the general information.
<br /><br /><b>List of prioritized small variants</b>
<table>
<tr><td><b>Variant</b></td><td><b>Genotype</b></td><td><b>Genes</b></td><td><b>Details</b></td><td><b>Class</b></td><td><b>Inheritance</b></td><td><b>gnomAD allele frequency<br />(control cohort)</b></td><td><b>RNA</b></td></tr>
<tr>
<td>
chr13:40793233&nbsp;T&nbsp;&gt;&nbsp;A</td><td>hom (comp-het)</td>
<td>SLC25A15
</td>
<td>SLC25A15:ENST00000338625.1/NM_014252:c.7T>A:p.Ser3Thr</td>
<td>3</td>
<td>n/a</td>
<td>n/a</td>
<td>splicing effect validated by RNA dataset</td>
</tr>
<tr><td colspan="8">OMIM ID: 603861 Details: GENE=SLC25A15 PHENOS=Hyperornithinemia-hyperammonemia-homocitrullinemia syndrome,238970,Autosomal recessive</td></tr>
<tr>
<td>
chr16:89531961&nbsp;G&nbsp;&gt;&nbsp;A</td><td>het (de-novo) (mosaic)</td>
<td>SPG7
</td>
<td>SPG7:ENST00000268704.1:c.1045G>A:p.Gly349Ser</td>
<td>5</td>
<td>n/a</td>
<td>0.080%</td>
<td>n/a</td>
</tr>
<tr><td colspan="8">OMIM ID: 602783 Details: GENE=SPG7 PHENOS=Spastic paraplegia 7,autosomal recessive,607259,Autosomal recessive,Autosomal dominant</td></tr>
</table>
<br /><b>List of prioritized copy-number variants and/or structural variants</b>
<table>
<tr><td><b>CNV/SV</b></td><td><b>Position</b></td><td><b>Size</b></td><td><b>copy-number/genotype</b></td><td><b>Genes</b></td><td><b>Class</b></td><td><b>Inheritance</b></td><td><b>RNA</b></td></tr>
<tr>
<td>deletion</td>
<td>chr21:10541092-10649856</td>
<td>108.765 kb / 22 regions</td>
<td>1 (comp-het)</td>
<td>IGHV1OR21-1, TPTE</td>
<td>n/a</td>
<td>n/a</td>
<td>no splicing effect found in RNA dataset</td>
</tr>
<tr>
<td>deletion</td>
<td>chr21:26799369-26991734</td>
<td>192.366 kb / 13 regions</td>
<td>0 (comp-het)</td>
<td></td>
<td>n/a</td>
<td>n/a</td>
<td>no splicing effect found in RNA dataset</td>
</tr>
<tr>
<td>insertion</td>
<td>chr1:934143-934144</td>
<td></td><td>hom</td>
<td>HES4</td>
<td>n/a</td>
<td>n/a</td>
<td>RNA dataset not usable</td>
</tr>
<tr>
<td>translocation</td>
<td>chr1:1584540-1584550 &lt;-&gt; chr19:2301860-2301870</td>
<td></td><td>hom</td>
<td></td>
<td>n/a</td>
<td>n/a</td>
<td>n/a</td>
</tr>
<tr>
<td>deletion</td>
<td>chr1:1598413-1598580</td>
<td>0.168 kb</td><td>hom</td>
<td>SLC35E2B</td>
<td>n/a</td>
<td>n/a</td>
<td>n/a</td>
</tr>
</table>
<p>&nbsp;</p>
<table>
<tr><td><b>variant type</b></td><td><b>regions</b></td><td><b>Genes</b></td><td><b>Inheritance</b></td><td><b>comment</b></td></tr>
<tr>
<td>uncalled CNV</td>
<td>chr2:123456-789012</td>
<td>EPRS</td>
<td>AR</td>
<td>This is a comment!
 And it has
 multiple lines!
</td>
</tr>
</table>

<p><b>Classification of variants:</b>
<br />Classification and interpretation of variants: The classification of variants is based on the criteria of Plon et al. (PMID: 18951446). A short description of each class can be found in the following
<br /><b>Class 5, pathogenic variant:</b> The variant is considered to be the cause of the patient's disease.
<br /><b>Class 4, probably pathogenic variants:</b> The identified variant is considered to be the probable cause of the patient's disease. This information should be used cautiously for clinical decision-making, as there is still a degree of uncertainty.
<br /><b>Class 3, variant of unclear significance (VUS):</b> The variant has characteristics of being an independent disease-causing mutation, but insufficient or conflicting evidence exists.
<br /><b>Class 2, most likely benign variants:</b> The variant is not likely to be the cause of the tested disease. Class 2 variants are not reported, but can be provided upon request.
<br /><b>Class 1, benign variants:</b> The variant is not considered to be the cause of the tested disease. Class 1 variants are not reported, but can be provided upon request.
</p>

<p><b>Coverage statistics of target region</b>
<br />Average sequencing depth: 125.47
<br />Average sequencing depth (chrMT): 0.00
</p>
<p><b>Gap report based on entire target region</b>
<br />Bases: 271536
<br />Percentage of regions with depth &lt;20: 15449
<br />Percentage gaps: 5.69%
<br />Genes without gaps: 
<br />Genes with incomplete coverage (missing bp in brackets): CYP7B1 <span style="font-size: 8pt;">(243)</span>, SLC25A15 <span style="font-size: 8pt;">(129)</span>, SPG7 <span style="font-size: 8pt;">(126)</span>
<p>Details regions with depth &lt;20:
</p>
<table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>
CYP7B1</td><td>243</td><td>chr8</td><td>64615713-64615740, 64616216-64616218, 64624393-64624515, 64798456-64798544
</td>
</tr>
<tr>
<td>
SLC25A15</td><td>129</td><td>chr13</td><td>40799186-40799314
</td>
</tr>
<tr>
<td>
SPG7</td><td>126</td><td>chr16</td><td>89524046-89524104, 89554476-89554542
</td>
</tr>
<tr>
<td>no gene overlap</td><td>14951</td><td>-</td><td>-</td>
</tr>
</table>
<br />Gaps closed by Sanger sequencing:<br /><table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>SPG7</td><td>59</td><td>chr16</td><td>89524046-89524104</td></tr>
</table>
Base sum:59<p>Gaps checked by visual inspection of raw data:<br /><table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>CYP7B1</td><td>28</td><td>chr8</td><td>64615713-64615740</td></tr>
<tr>
<td>CYP7B1</td><td>3</td><td>chr8</td><td>64616216-64616218</td></tr>
</table>
Base sum:31</p>
<p>gaps remaining
</p>
<table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>
CYP7B1</td><td>212</td><td>chr8</td><td>64624393-64624515, 64798456-64798544
</td>
</tr>
<tr>
<td>
SLC25A15</td><td>129</td><td>chr13</td><td>40799186-40799314
</td>
</tr>
<tr>
<td>
SPG7</td><td>67</td><td>chr16</td><td>89554476-89554542
</td>
</tr>
<tr>
<td>no gene overlap</td><td>14951</td><td>-</td><td>-</td>
</tr>
</table>
<p>Percentage of regions with depth &lt;20 after closing gaps: 15359
<br />Percentage gaps after closing gaps: 5.66%
</p></p>
<p><b>Gap report based on exons of target region &#177; 20 Bases</b>
<br />Bases: 5975
<br />Percentage of regions with depth &lt;20: 528
<br />Percentage gaps: 8.84%
<br />Genes without gaps: 
<br />Genes with incomplete coverage (missing bp in brackets): CYP7B1 <span style="font-size: 8pt;">(263)</span>, SLC25A15 <span style="font-size: 8pt;">(129)</span>, SPG7 <span style="font-size: 8pt;">(136)</span>
<p>Details regions with depth &lt;20:
</p>
<table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>
CYP7B1</td><td>263</td><td>chr8</td><td>64615713-64615740, 64616216-64616218, 64624383-64624515, 64798446-64798544
</td>
</tr>
<tr>
<td>
SLC25A15</td><td>129</td><td>chr13</td><td>40799186-40799314
</td>
</tr>
<tr>
<td>
SPG7</td><td>136</td><td>chr16</td><td>89524046-89524104, 89554466-89554542
</td>
</tr>
</table>
<br />Gaps closed by Sanger sequencing:<br /><table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>SPG7</td><td>59</td><td>chr16</td><td>89524046-89524104</td></tr>
</table>
Base sum:59<p>Gaps checked by visual inspection of raw data:<br /><table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>CYP7B1</td><td>28</td><td>chr8</td><td>64615713-64615740</td></tr>
<tr>
<td>CYP7B1</td><td>3</td><td>chr8</td><td>64616216-64616218</td></tr>
</table>
Base sum:31</p>
<p>gaps remaining
</p>
<table>
<tr><td><b>Gene</b></td><td><b>Bases</b></td><td><b>Chromosome</b></td><td><b>Coordinates (hg38)</b></td></tr>
<tr>
<td>
CYP7B1</td><td>232</td><td>chr8</td><td>64624383-64624515, 64798446-64798544
</td>
</tr>
<tr>
<td>
SLC25A15</td><td>129</td><td>chr13</td><td>40799186-40799314
</td>
</tr>
<tr>
<td>
SPG7</td><td>77</td><td>chr16</td><td>89554466-89554542
</td>
</tr>
</table>
<p>Percentage of regions with depth &lt;20 after closing gaps: 438
<br />Percentage gaps after closing gaps: 7.33%
</p></p>

<p><b>Coverage statistics of RNA sample</b>
<br />Number of reads: 10.99 Mio
<br />Average sequencing depth: 15.85
<br />Average sequencing depth of housekeeping genes: 263.87
<br />Covered genes: 136985
</p>

<p><b>OMIM gene and phenotypes</b>
</p>
<table>
<tr><td><b>Gene</b></td><td><b>gene MIM</b></td><td><b>phenotype MIM</b></td><td><b>phenotype</b></td><td><b>preferred phenotype</b></td>
</tr><tr><td>CYP7B1</td><td>603711</td><td>270800</td><td>Spastic paraplegia 5A, autosomal recessive, 270800 (3)</td><td>yes</td>
</tr><tr><td>SLC25A15</td><td>603861</td><td>238970</td><td>Hyperornithinemia-hyperammonemia-homocitrullinemia syndrome, 238970 (3)</td><td>no</td>
</tr><tr><td>SPG7</td><td>602783</td><td>607259</td><td>Spastic paraplegia 7, autosomal recessive, 607259 (3)</td><td>no</td>
</tr></table>

<p><b>Polygenic Risk Scores (PRS)</b></p>
<table>
<tr><td><b>Trait</b></td><td><b>Publication</b></td><td><b>Score</b></td><td><b>z-score</b></td><td><b>population (estimated from NGS)</b></td></tr>
<tr><td>Breast Cancer</td><td>Mavaddat N et al. Am J Hum Genet (2018). doi:10.1016/j.ajhg.2018.11.002</td><td>-0.2773</td><td>0.240</td><td>European</td></tr></table>
<p>A validated risk estimation program must be used to judge the clinical importance of a PRS, e.g. CanRisk.org for breast cancer. The ethnicity of the patient must also be considered.</p>
	</body>
</html>
